### PR TITLE
Use server capabilities

### PR DESCRIPTION
**Pull request by rullzer**
Replaces #2943
Should fix: #2815

By making use of the capabilities provided by the server we can provide a better UX.

1. Do not allow a user to remove expire date/password if the admin enforces it
2. Require a password from the user if this is enforced.

### CODE REVIEW

**Comment by dragotin** [comment]
Two questions/suggestions:
- I think we should have a Capabilities object that with proper methods, so that it reads:
```if( account()->capabilities()->pubLinkForcePassword() ) { ...``` instead of the map stuff we have now.
- Please do not use the QInputDialog. The dialog has a password field already, which should be made visible rather.

What do you think?

**Comment by rullzer** [comment]
> - I think we should have a Capabilities object that with proper methods, so that it reads:
```if( account()->capabilities()->pubLinkForcePassword() ) { ...``` instead of the map stuff we have now.

I agree. Now the questions is where and how to do this. Do we do this on a as needed basis... so I create a shareCapabilities class where all that is properly parsed. Or do we do it in 1 place where we parse everything?

> - Please do not use the QInputDialog. The dialog has a password field already, which should be made visible rather.

Yes, this should be changed indeed.

**Comment by rullzer** [comment]
@dragotin both your comments are address in the latest two commits...
Code should be better readable now and no annoying extra inputdialog anymore.

File: src/libsync/account.h [hunk 1]
```diff
@@ -170,7 +171,7 @@ protected Q_SLOTS:
     QUrl _url;
     QList<QSslCertificate> _approvedCerts;
     QSslConfiguration _sslConfiguration;
-    QVariantMap _capabilities;
+    Capabilities *_capabilities;
```
Comment (ogoffart) [inline src/libsync/account.h]: This should not be a pointer.  Just make it a normal object like the QVariantMap was.

File: src/libsync/account.h [hunk 2]
```diff
@@ -139,7 +140,7 @@ class OWNCLOUDSYNC_EXPORT Account : public QObject {
     void setCertificate(const QByteArray certficate = QByteArray(), const QString privateKey = QString());
 
     void setCapabilities(const QVariantMap &caps);
-    QVariantMap capabilities();
+    Capabilities * capabilities();
```
Comment (ogoffart) [inline src/libsync/account.h]: return a const reference

Comment (ogoffart) [inline src/libsync/account.h]: and make const  (it should have been const before)

**Comment by rullzer** [comment]
@ogoffart thanks and fixed!

File: src/libsync/capabilities.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ * server
+ * @ingroup libsync
+ */
+class Capabilities {
+
+public:
+    Capabilities(const Capabilities& capabilities);
+    Capabilities(const QVariantMap capabilities);
+
+    Capabilities operator=(Capabilities rhs)
```
Comment (ogoffart) [inline src/libsync/capabilities.h]: This looks wrong, you are not assinging anything.

The default generated copy constructor and operator are fine.  Implementing them breaks the move sementic.  (see "Rule of zero")

Comment (rullzer) [inline src/libsync/capabilities.h]: Aaah you are right...

**Pull request merged**